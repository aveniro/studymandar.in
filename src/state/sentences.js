import {createStore, createEvent} from 'effector';

export const sentencesState = createStore([]);

export const sentencesApi = {
    addSentence: createEvent(),
    updateSentences: createEvent()
};

sentencesState.on(sentencesApi.addSentence, (oldSentencesState, newSentence) => [...oldSentencesState, newSentence]);

sentencesState.on(sentencesApi.updateSentences, (oldSentencesState, newSentence) => newSentence);