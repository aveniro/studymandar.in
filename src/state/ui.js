import {createApi, createStore} from 'effector';

export const menuStore = createStore('closed');

export const menuApi = createApi(menuStore, {
    toggle: (menuStore) => menuStore === 'open' ? 'closed' : 'open',
    close: () => 'closed'
});

export const loadStore = createStore('inactive');

export const loadApi = createApi(loadStore, {
    start: () => 'active',
    stop: () => 'inactive'
});