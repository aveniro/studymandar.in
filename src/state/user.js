import {createEvent, createStore} from 'effector';

export const userState = createStore(null);

export const userApi = {
    updateUser: createEvent(),
    requestLoginRegister: createEvent(),
    rejectLoginRegister: createEvent()
};

userState.on(userApi.updateUser, (userState, newState) => newState);