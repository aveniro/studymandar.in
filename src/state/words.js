import {createStore, createEvent} from 'effector';

export const wordsState = createStore([]);

export const wordsApi = {
    updateWord: createEvent(),
    addWord: createEvent(),
    deleteWord: createEvent(),
    ocrCorrect: createEvent(),
    ocrIncorrect: createEvent(),
    vocabCorrect: createEvent(),
    vocabIncorrect: createEvent(),
    updateWords: createEvent()
};

wordsState.on(wordsApi.updateWord, (oldWordsState, word) => {
    const index = oldWordsState.findIndex(el => el.id === word.id);
    if(index !== -1) {
        oldWordsState[index] = word;
    } else {
        oldWordsState.push(word);
    }

    return [...oldWordsState];
});

wordsState.on(wordsApi.addWord, (oldWordsState, newWord) => [...oldWordsState, newWord]);

wordsState.on(wordsApi.deleteWord, (oldWordsState, word) => {
    const index = oldWordsState.findIndex(el => el.id === word.id);

    if(index !== -1) {
        oldWordsState.splice(index, 1);
    }
    
    return [...oldWordsState];
});

wordsState.on(wordsApi.ocrCorrect, (oldWordsState, word) => {
    ++word.ocrTested || (word.ocrTested = 1);
    ++word.ocrCorrect || (word.ocrCorrect = 1);

    word.ocrRatio = (word.ocrTested > 0 ? word.ocrCorrect / word.ocrTested : 0).toFixed(2);

    wordsApi.updateWord(word);
});

wordsState.on(wordsApi.ocrIncorrect, (oldWordsState, word) => {
    ++word.ocrTested || (word.ocrTested = 1);
    word.ocrCorrect || (word.ocrCorrect = 0);

    word.ocrRatio = (word.ocrTested > 0 ? word.ocrCorrect / word.ocrTested : 0).toFixed(2);

    wordsApi.updateWord(word);
});

wordsState.on(wordsApi.vocabCorrect, (oldWordsState, word) => {
    ++word.vocabTested || (word.vocabTested = 1);
    ++word.vocabCorrect || (word.vocabCorrect = 1);

    word.vocabRatio = (word.vocabTested > 0 ? word.vocabCorrect / word.vocabTested : 0).toFixed(2);

    wordsApi.updateWord(word);
});

wordsState.on(wordsApi.vocabIncorrect, (oldWordsState, word) => {
    ++word.vocabTested || (word.vocabTested = 1);
    word.vocabCorrect || (word.vocabCorrect = 0);

    word.vocabRatio = (word.vocabTested > 0 ? word.vocabCorrect / word.vocabTested : 0).toFixed(2);

    wordsApi.updateWord(word);
});

wordsState.on(wordsApi.updateWords, (oldWordsState, newWordsState) => newWordsState);