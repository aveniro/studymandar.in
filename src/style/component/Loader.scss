.loader {
    z-index: 10000000;
    position: absolute;
    height: 100vh;
    width: 100vw;
    opacity: 0;
    transition: 200ms;
    display: flex;
    justify-content: center;
    align-items: center;

    background: #3891A6;

    svg {
        animation: spin 750ms ease-in-out infinite;

        @keyframes spin {
            from {
                transform: rotateZ(0deg);
            }

            to {
                transform: rotateZ(360deg);
            }
        }
    }

    &[data-status="active"] {
        opacity: 1;
    }

    &[data-status="inactive"] {
        pointer-events: none;
    }
}