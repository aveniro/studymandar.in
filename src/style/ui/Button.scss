@import '#/common';

.button {
    cursor: pointer;
    padding: 0.5em;
    color: white;
    border-radius: 0.2em;
    background: $button-color;
    text-align: center;
    transition: 300ms;
    user-select: none;
    display: flex;
    flex-direction: row-reverse;
    justify-content: center;
    align-items: center;

    &:hover {
        background: $button-color-hover;
    }

    &:active {
        background: $button-color-active;
    }

    $svgsize: 1em;
    $svgpad: 0.5em;

    > svg {
        width: $svgsize;
        height: $svgsize;
        margin-right: $svgpad;
        transition: 300ms;
    }

    &[data-status="loading"] {
        > svg { 
            animation: spin 750ms ease-in-out infinite;

            @keyframes spin {
                from {
                    transform: rotateZ(0deg);
                }

                to {
                    transform: rotateZ(360deg);
                }
            }
        }
    }

    &[data-status="not-loading"] {
        > svg {
            opacity: 0;
            width: 0;
            padding-right: 0;
        }
    }
}