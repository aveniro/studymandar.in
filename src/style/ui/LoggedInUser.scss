.logged-in-user-container {
    position: relative;

    .logged-in-user {
        display: flex;
        flex-direction: row;
        align-items: center;
        border: 1px solid rgba(0,0,0,0.2);
        padding: 0.5em 1em 0.5em 1em;
        border-radius: 0.25em;
        cursor: pointer;
    
        box-shadow: 2px 2px 4px 1px rgba(0,0,0,0.05);
    
        transition: 300ms;
    
        &:hover {
            background: rgba(0,0,0,0.02);
        }
    
        &:active {
            background: rgba(0,0,0,0.05);
        }
        
        > img {
            border-radius: 100%;
            margin-left: 1em;
            width: 1.5em;
            height: 1.5em;
            user-select: none;
            object-fit: cover;
        }
    
        > .user-email {
            font-size: 0.8em;
            user-select: none;
        }
    }

    > .open-menu-thingy {
        position: absolute;
        top: 100%;
        left: 0;
        right: 0;
        background: rgb(236, 236, 236);
        transform: scaleY(0);
        transition: 300ms;
        transform-origin: top center;

        > .drop-menu-item {
            text-align: left;
            padding: 0.5em;
            font-size: 0.8em;
            cursor: pointer;
        }

        &[data-status="open"] {
            transform: scaleY(1);
        }
    }
}