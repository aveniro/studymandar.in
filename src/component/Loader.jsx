import {h}        from 'preact';
import {useState} from 'preact/hooks';
import {useStore} from 'effector-react';

import {loadStore} from 'state/ui';

import loadingMessages from '@/loadingMessages.json';

import '#/component/Loader.scss';

export default function Loader() {
    const loadState = useStore(loadStore);
    const [message] = useState(loadingMessages[Math.floor(Math.random() * loadingMessages.length)]);

    return (
        <div data-status={loadState} className="loader">
            <div className="load-box">
                <svg width="50" height="50" viewBox="0 0 50 50">
                    <path fill="#FFFFFF" d="M25,5A20.14,20.14,0,0,1,45,22.88a2.51,2.51,0,0,0,2.49,2.26h0A2.52,2.52,0,0,0,50,22.33a25.14,25.14,0,0,0-50,0,2.52,2.52,0,0,0,2.5,2.81h0A2.51,2.51,0,0,0,5,22.88,20.14,20.14,0,0,1,25,5Z"></path>
                </svg>
            </div>
            <div className="message-area">
                {message}
            </div>
        </div>
    );
}