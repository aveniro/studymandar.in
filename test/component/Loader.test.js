import {h}     from 'preact';
import Enzyme  from 'enzyme';
import Adapter from 'enzyme-adapter-preact-pure';
import Loader  from '@/component/Loader';

import {loadApi} from 'state/ui';

Enzyme.configure({ adapter: new Adapter() });

describe('<Loader /> component', () => {
    const container = Enzyme.shallow(<Loader />);

    it('should match the snapshot', () => {
        expect(container.html()).toMatchSnapshot();
    });

    it('should stop loading', () => {
        loadApi.stop();
        expect(container.html()).toMatchSnapshot();
    });

    it('should start loading', () => {
        loadApi.start();
        expect(container.html()).toMatchSnapshot();
    });
});