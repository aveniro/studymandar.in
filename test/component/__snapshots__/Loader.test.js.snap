// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loader /> component should match the snapshot 1`] = `"<div data-status=\\"inactive\\" class=\\"loader\\"><svg width=\\"50\\" height=\\"50\\" viewBox=\\"0 0 50 50\\"><path fill=\\"#FFFFFF\\" d=\\"M25,5A20.14,20.14,0,0,1,45,22.88a2.51,2.51,0,0,0,2.49,2.26h0A2.52,2.52,0,0,0,50,22.33a25.14,25.14,0,0,0-50,0,2.52,2.52,0,0,0,2.5,2.81h0A2.51,2.51,0,0,0,5,22.88,20.14,20.14,0,0,1,25,5Z\\"></path></svg></div>"`;

exports[`<Loader /> component should start loading 1`] = `"<div data-status=\\"active\\" class=\\"loader\\"><svg width=\\"50\\" height=\\"50\\" viewBox=\\"0 0 50 50\\"><path fill=\\"#FFFFFF\\" d=\\"M25,5A20.14,20.14,0,0,1,45,22.88a2.51,2.51,0,0,0,2.49,2.26h0A2.52,2.52,0,0,0,50,22.33a25.14,25.14,0,0,0-50,0,2.52,2.52,0,0,0,2.5,2.81h0A2.51,2.51,0,0,0,5,22.88,20.14,20.14,0,0,1,25,5Z\\"></path></svg></div>"`;

exports[`<Loader /> component should stop loading 1`] = `"<div data-status=\\"inactive\\" class=\\"loader\\"><svg width=\\"50\\" height=\\"50\\" viewBox=\\"0 0 50 50\\"><path fill=\\"#FFFFFF\\" d=\\"M25,5A20.14,20.14,0,0,1,45,22.88a2.51,2.51,0,0,0,2.49,2.26h0A2.52,2.52,0,0,0,50,22.33a25.14,25.14,0,0,0-50,0,2.52,2.52,0,0,0,2.5,2.81h0A2.51,2.51,0,0,0,5,22.88,20.14,20.14,0,0,1,25,5Z\\"></path></svg></div>"`;
